<a href="https://colab.research.google.com/github/Rahaf-altuwairesh/Data-Analysis-and-Machine-Learning-STC-Jawwy-dataset-/blob/main/Task_1_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# STC Jawwy

## Analysis of Video Views by Resolution

In this analysis, we present the percentage of video views based on video resolution (SD and HD) using a pie chart created with Plotly Express. This visualization aims to illustrate the distribution of video views between the different resolutions and understand user preferences.

### Key Points in the Analysis:
**1. Calculation of the Percentage of Views for Each Resolution**:
- The total number of views for all resolutions was calculated.
- The percentage for each resolution was calculated by dividing the number of views for each resolution by the total number of views and multiplying the result by 100 to obtain the percentage.

**2. Creating a DataFrame for Plotly**:
- A DataFrame was created containing the resolutions (SD and HD) and the percentages of views for each resolution.

**3. Visualizing the Data Using a Pie Chart**:
- The pie chart was created using the Plotly Express library.
- The colors of the chart were customized to be more serious, using green (#2e8b57) for SD resolution and blue (#4682b4) for HD resolution.
- Displaying the percentage and label inside the pie chart slices was enabled for better clarity of information.

In [1]:
"""
Here we install libraries that are not installed by default
Example:  pyslsb
Feel free to add any library you are planning to use.
"""
!pip install pyxlsb

In [2]:
# Import the required libraries
"""
Please feel free to import any required libraries as per your needs
"""
import pandas as pd     # provides high-performance, easy to use structures and data analysis tools
import pyxlsb           # Excel extention to read xlsb files (the input file)
import numpy as np      # provides fast mathematical computation on arrays and matrices

# Jawwy dataset
The dataset consists of meta details about the movies and tv shows as genre.
Also details about Users activities, spent duration and if watching in High definition or standard definition.
You have to analyse this dataset to find top insights, findings and to solve the four tasks assigned to you.

In [9]:
from google.colab import files
import pandas as pd
import io

# Upload the file
uploaded = files.upload()

# Iterate over the uploaded files (in case multiple files are uploaded)
for filename in uploaded.keys():
    # Read the first sheet in the Excel file
    dataframe = pd.read_excel(io.BytesIO(uploaded[filename]), sheet_name=0)  # 0 refers to the first sheet
    print(dataframe.head())  # Display the first few rows of the dataframe


Saving stc TV Data Set_T1.xlsb to stc TV Data Set_T1 (1).xlsb
   Column1  date_  user_id_maped           program_name  duration_seconds  \
0        1  42882          26138             100 treets                40   
1        3  42876           7946                 Moana                 17   
2        4  42957           7418  The Mermaid Princess                  8   
3        5  42942          19307  The Mermaid Princess                 76   
4        7  42923          15860             Churchill                 87   

  program_class  season  episode                              program_desc  \
0         MOVIE       0        0                    Drama Movie100 Streets   
1         MOVIE       0        0                 Animation MovieMoana (HD)   
2         MOVIE       0        0  Animation MovieThe Mermaid Princess (HD)   
3         MOVIE       0        0  Animation MovieThe Mermaid Princess (HD)   
4         MOVIE       0        0             Biography MovieChurchill (HD)   

  prog

In [10]:
# check the data shape
dataframe.shape

(1048575, 13)

In [11]:
# display the first 5 rows
dataframe.head()

Column1  date_  user_id_maped           program_name  duration_seconds  \
0        1  42882          26138             100 treets                40   
1        3  42876           7946                 Moana                 17   
2        4  42957           7418  The Mermaid Princess                  8   
3        5  42942          19307  The Mermaid Princess                 76   
4        7  42923          15860             Churchill                 87   

  program_class  season  episode                              program_desc  \
0         MOVIE       0        0                    Drama Movie100 Streets   
1         MOVIE       0        0                 Animation MovieMoana (HD)   
2         MOVIE       0        0  Animation MovieThe Mermaid Princess (HD)   
3         MOVIE       0        0  Animation MovieThe Mermaid Princess (HD)   
4         MOVIE       0        0             Biography MovieChurchill (HD)   

  program_genre  series_title  hd         original_name  
0         Drama             0   0            100 treets  
1     Animation             0   1                 Moana  
2     Animation             0   1  The Mermaid Princess  
3     Animation             0   1  The Mermaid Princess  
4     Biography             0   1             Churchill

In [12]:
# Data Preprocessing on the input data
dataframe = dataframe.drop(columns=['Column1']) # dropping the index column
dataframe['program_name'] = dataframe['program_name'].str.strip()  # trim spaces in movies names to avoid misspellings in input data
dataframe[['duration_seconds', 'season','episode','series_title','hd']] = dataframe[['duration_seconds', 'season','episode','series_title','hd']].apply(pd.to_numeric)  # read numeric columns as numeric data types
dataframe[['user_id_maped', 'program_name','program_class','program_desc','program_genre','original_name']] = dataframe[['user_id_maped', 'program_name','program_class','program_desc','program_genre','original_name']].astype(str) # read string columns as string data types

In [13]:
# display the dataset after applying data types
dataframe.head(30)

date_ user_id_maped                     program_name  duration_seconds  \
0   42882         26138                       100 treets                40   
1   42876          7946                            Moana                17   
2   42957          7418             The Mermaid Princess                 8   
3   42942         19307             The Mermaid Princess                76   
4   42923         15860                        Churchill                87   
5   42966         20775  Beavis And Butt-Head Do America                 3   
6   42912         11374             The Mermaid Princess             33577   
7   43188          6358                             Coco                14   
8   43127         11660                           Kidnap                85   
9   42824          5155                   The Accountant                42   
10  42835         30889            The Birth of a Nation                71   
11  42977          1035            The Amazing pider-Man                85   
12  43133         12598                         Baywatch                46   
13  43002          7805                  Alien: Covenant                90   
14  42998         33846                    Going in tyle                86   
15  43120         12333      Kingsman: The Golden Circle               153   
16  43141         25343    Billy Elliot the Musical Live                 2   
17  42948         30971                    Going in tyle                35   
18  42911         26546                            Moana                43   
19  42909         20306                            Moana                12   
20  42960         16962                    Pitch Perfect                28   
21  42820         16880            Surf's Up : WaveMania                84   
22  43072         19694  Hell On Wheels  The White pirit              2650   
23  42870         22529                            Sully                69   
24  42922         21917                             Goat                12   
25  42986         28952                     Last Knights                47   
26  42977         25404                 Land Of The Dead              3553   
27  43164         14270                         Geostorm                 2   
28  42857         14249                           Trolls                20   
29  43007         21019             The Mermaid Princess                21   

      program_class  season  episode  \
0             MOVIE       0        0   
1             MOVIE       0        0   
2             MOVIE       0        0   
3             MOVIE       0        0   
4             MOVIE       0        0   
5             MOVIE       0        0   
6             MOVIE       0        0   
7             MOVIE       0        0   
8             MOVIE       0        0   
9             MOVIE       0        0   
10            MOVIE       0        0   
11            MOVIE       0        0   
12            MOVIE       0        0   
13            MOVIE       0        0   
14            MOVIE       0        0   
15            MOVIE       0        0   
16            MOVIE       0        0   
17            MOVIE       0        0   
18            MOVIE       0        0   
19            MOVIE       0        0   
20            MOVIE       0        0   
21            MOVIE       0        0   
22  SERIES/EPISODES       2        7   
23            MOVIE       0        0   
24            MOVIE       0        0   
25            MOVIE       0        0   
26            MOVIE       0        0   
27            MOVIE       0        0   
28            MOVIE       0        0   
29            MOVIE       0        0   

                                      program_desc program_genre  \
0                           Drama Movie100 Streets         Drama   
1                        Animation MovieMoana (HD)     Animation   
2         Animation MovieThe Mermaid Princess (HD)     Animation   
3         Animation MovieThe Mermaid Princess (HD)     Animation   
4   

In [14]:
# describe the numeric values in the dataset
dataframe.describe()

date_  duration_seconds        season       episode  \
count  1.048575e+06      1.048575e+06  1.048575e+06  1.048575e+06   
mean   4.301202e+04      1.230957e+03  1.342139e+00  6.157952e+00   
std    1.242834e+02      6.821058e+03  2.104095e+00  1.222015e+01   
min    4.280800e+04      2.000000e+00  0.000000e+00  0.000000e+00   
25%    4.289600e+04      5.200000e+01  0.000000e+00  0.000000e+00   
50%    4.302200e+04      1.190000e+02  1.000000e+00  1.000000e+00   
75%    4.312100e+04      1.328000e+03  1.000000e+00  9.000000e+00   
max    4.322000e+04      1.461329e+06  2.300000e+01  2.820000e+02   

       series_title            hd  
count  1.048575e+06  1.048575e+06  
mean   1.205922e-02  3.862728e-01  
std    1.091504e-01  4.868946e-01  
min    0.000000e+00  0.000000e+00  
25%    0.000000e+00  0.000000e+00  
50%    0.000000e+00  0.000000e+00  
75%    0.000000e+00  1.000000e+00  
max    1.000000e+00  1.000000e+00

In [15]:
# check if any column has null value in the dataset
dataframe.isnull().any()

date_               False
user_id_maped       False
program_name        False
duration_seconds    False
program_class       False
season              False
episode             False
program_desc        False
program_genre       False
series_title        False
hd                  False
original_name       False
dtype: bool

In [16]:
#Check the data type
dataframe.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1048575 entries, 0 to 1048574
Data columns (total 12 columns):
 #   Column            Non-Null Count    Dtype 
---  ------            --------------    ----- 
 0   date_             1048575 non-null  int64 
 1   user_id_maped     1048575 non-null  object
 2   program_name      1048575 non-null  object
 3   duration_seconds  1048575 non-null  int64 
 4   program_class     1048575 non-null  object
 5   season            1048575 non-null  int64 
 6   episode           1048575 non-null  int64 
 7   program_desc      1048575 non-null  object
 8   program_genre     1048575 non-null  object
 9   series_title      1048575 non-null  int64 
 10  hd                1048575 non-null  int64 
 11  original_name     1048575 non-null  object
dtypes: int64(6), object(6)
memory usage: 96.0+ MB


In [17]:
#Check the resolution type
resolution_counts = dataframe['hd'].value_counts()

# Display the counts
print(resolution_counts)


hd
0    643539
1    405036
Name: count, dtype: int64


In [18]:
# we import Visualization libraries
# you can ignore and use any other graphing libraries
import matplotlib.pyplot as plt # a comprehensive library for creating static, animated, and interactive visualizations
import plotly #a graphing library makes interactive, publication-quality graphs. Examples of how to make line plots, scatter plots, area charts, bar charts, error bars, box plots, histograms, heatmaps, subplots, multiple-axes, polar charts, and bubble charts.
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots

In [19]:
# Calculate the percentage of views for each resolution
total_views = resolution_counts.sum()
resolution_percentages = (resolution_counts / total_views) * 100

# Create a DataFrame for Plotly
resolution_data = pd.DataFrame({
    'Resolution': ['SD', 'HD'],
    'Percentage': [resolution_percentages.get(0, 0), resolution_percentages.get(1, 0)]
})

# Plotting the pie chart using Plotly Express
fig = px.pie(
    resolution_data,
    names='Resolution',
    values='Percentage',
    title='Percentage of Views by Video Resolution',
    color='Resolution',
    color_discrete_sequence=['#2e8b57', '#4682b4'],  # Customize colors as needed
    labels={'Percentage': 'Percentage of Views'}
)

# Display the plot
fig.update_traces(textinfo='percent+label')  # Show percentage and label in the pie slices
fig.show()

## Conclusion:

**This visualization shows that the majority of video views were in SD resolution, accounting for 61.4% of total views. Meanwhile, views in HD resolution accounted for 38.6% of the total. This may indicate that users prefer watching videos in SD resolution more than HD, possibly due to preferences for video quality or limitations related to internet speeds or devices used.**

# Task 1
##### You are required to work on task one to study and HD flag for available dataset

In [20]:
# make a copy of the dataframe for working on task 1
df=dataframe.copy()

In [21]:
# Here we try to get the most watched movies (Total Views / Total Users Views / Total watch time)
# For series we concatenated the Session episode to differentiate between episodes
grouped=df.copy()
grouped.loc[grouped['program_class'] == 'SERIES/EPISODES', 'program_name'] = grouped['program_name']+'_SE'+grouped['season'].astype(str)+'_EP'+grouped['episode'].astype(str)
grouped = grouped.groupby(['program_name','program_class'])\
.agg({'user_id_maped': [('co1', 'nunique'),('co2', 'count')],\
      'duration_seconds': [('co3', 'sum')] }).reset_index()
grouped.columns = ['program_name','program_class','No of Users who Watched', 'No of watches', 'Total watch time in seconds']
grouped['Total watch time in houres']=grouped['Total watch time in seconds']/3600
grouped = grouped.drop(columns=['Total watch time in seconds'])
grouped = grouped.sort_values(by=['Total watch time in houres', 'No of watches','No of Users who Watched'], ascending=False).reset_index(drop=True)


In [22]:
# show the result
grouped.head(35)

program_name    program_class  \
0                              The Boss Baby            MOVIE   
1                      The Amazing pider-Man            MOVIE   
2                            The Expendables            MOVIE   
3                                      Moana            MOVIE   
4                                     Trolls            MOVIE   
5                                       Bean            MOVIE   
6                                  The murfs            MOVIE   
7                         Hotel Transylvania            MOVIE   
8          Cloudy With a Chance of Meatballs            MOVIE   
9                The Man With The Iron Fists            MOVIE   
10                                      Salt            MOVIE   
11                                  Unbroken            MOVIE   
12                                ParaNorman            MOVIE   
13                        Youm Maloosh Lazma            MOVIE   
14                                 Ferdinand            MOVIE   
15                              White Chicks            MOVIE   
16                             Jurassic Park            MOVIE   
17                          The November Man            MOVIE   
18                              Total Recall            MOVIE   
19                                Robin Hood            MOVIE   
20                            Public Enemies            MOVIE   
21                            Daddy Day Camp            MOVIE   
22                                  Oblivion            MOVIE   
23                                     Blitz            MOVIE   
24            War for the Planet of the Apes            MOVIE   
25                                Inside Man            MOVIE   
26                                  Bad Boys            MOVIE   
27                                    Easy A            MOVIE   
28                                Battleship            MOVIE   
29                                  Baywatch            MOVIE   
30                               Police tory            MOVIE   
31  Blindspot     Erase Weary Youth_SE1_EP13  SERIES/EPISODES   
32                                  47 Ronin            MOVIE   
33                           Alien: Covenant            MOVIE   
34                      The Mermaid Princess            MOVIE   

    No of Users who Watched  No of watches  Total watch time in houres  
0                      3389          24047                 2961.350833  
1                      1011           2877                 1966.119167  
2                       853           2119                 1961.159444  
3                      2173           8081                 1706.176944  
4                      2613          13793                 1601.023056  
5                       949           3617                 1423.955000  
6                       867           3132                 1342.141111  
7                       491           1947                 1096.533611  
8                       683           2076                  948.674722  
9                       707           2505                  859.626389  
10                      563           1082                  767.392778  
11                      625           1429                  763.078333  
12                      614           1746                  747.065556  
13                     1131           2278                  718.109722  
14                     1278           6817                  714.223056  
15                      307            916                  711.840833  
16                      504           1192                  693.394444  
17                      494           1219                  679.492222  
18                      587           1108                  661.820000  
19                      588           1209                  643.935000  
20                      368            716                  634.035000  
21                      263            647                  625.338333  
22          

In [23]:
# plot top 10 Programs
fig = px.pie(grouped.head(10), values='Total watch time in houres', names='program_name',\
             hover_data=['program_class'],title='top 10 programs in total watch time in houres')
fig.show()

In [24]:
# Here we try to study the customer experience against Program class
grouped=df.copy()
grouped = grouped.groupby('program_class')\
.agg({'user_id_maped': [('co1', 'nunique'),('co2', 'count')],\
      'duration_seconds': [('co3', 'sum')] }).reset_index()
grouped.columns = ['program_class','No of Users who Watched', 'No of watches', 'Total watch time in seconds']
grouped['Total watch time in houres']=grouped['Total watch time in seconds']/3600
grouped = grouped.drop(columns=['Total watch time in seconds'])
grouped = grouped.sort_values(by=['Total watch time in houres', 'No of watches','No of Users who Watched'], ascending=False).reset_index(drop=True)


In [25]:
# show the result
grouped.head()

program_class  No of Users who Watched  No of watches  \
0  SERIES/EPISODES                     3901         560174   
1            MOVIE                    11355         488401   

   Total watch time in houres  
0               255097.787500  
1               103444.145556

In [26]:
# plot the total watch time against total number of users and report your findings
fig = px.pie(grouped, values='Total watch time in houres', names='program_class',\
             hover_data=['program_class'],title='Total duration spent by program_class')
fig2 = px.pie(grouped, values='No of Users who Watched', names='program_class',\
             hover_data=['program_class'],title='Total Users watching by program_class')

fig.update_traces(sort=False)
fig2.update_traces(sort=False)
fig.show()
fig2.show()

# Task 2

**TO DO study the relation and user's behaviour Against HD flag, feel free to include any comparison.**


In [27]:
# Create a copy of the original DataFrame to avoid modifying the original data
grouped = df.copy()

# Group the data by 'hd' and 'program_class'
# Aggregate the following:
# 1. The unique count of users who watched (nunique on user_id_maped)
# 2. The total count of watches (count on user_id_maped)
# 3. The sum of watch durations (sum on duration_seconds)
grouped = grouped.groupby(['hd', 'program_class']).agg({
    'user_id_maped': [('No of Users who Watched', 'nunique'),  # Unique users count
                      ('No of watches', 'count')],  # Total watch count
    'duration_seconds': [('Total watch time in seconds', 'sum')]  # Sum of watch duration in seconds
}).reset_index()

# Rename columns to be more readable after aggregation
grouped.columns = ['hd', 'program_class', 'No of Users who Watched', 'No of watches', 'Total watch time in seconds']

# Add a new column to convert total watch time from seconds to hours
grouped['Total watch time in hours'] = grouped['Total watch time in seconds'] / 3600

# Drop the "Total watch time in seconds" column as it is no longer needed
grouped = grouped.drop(columns=['Total watch time in seconds'])

# Sort the data by total watch time in hours, then by the number of watches, and finally by the number of users
grouped = grouped.sort_values(by=['Total watch time in hours', 'No of watches', 'No of Users who Watched'],
                              ascending=False).reset_index(drop=True)

In [28]:
# show the result
grouped.head()

hd    program_class  No of Users who Watched  No of watches  \
0   0  SERIES/EPISODES                     3282         486884   
1   1            MOVIE                    10880         331746   
2   0            MOVIE                     6093         156655   
3   1  SERIES/EPISODES                     2625          73290   

   Total watch time in hours  
0              229776.593333  
1               64856.366111  
2               38587.779444  
3               25321.194167

In [29]:
import plotly.graph_objects as go
from plotly.subplots import make_subplots

# Create subplots: 1 row and 2 columns, both of 'domain' type (for pie charts)
fig = make_subplots(1, 2, specs=[[{'type': 'domain'}, {'type': 'domain'}]],
                    subplot_titles=['SERIES/EPISODES', 'MOVIE'])

# Add pie chart for SERIES/EPISODES
fig.add_trace(go.Pie(labels=grouped[grouped['program_class'] == 'SERIES/EPISODES']['hd'],
                     values=grouped[grouped['program_class'] == 'SERIES/EPISODES']['No of Users who Watched'],
                     name='SERIES/EPISODES', direction='clockwise', sort=True),
              row=1, col=1)

# Add pie chart for MOVIE
fig.add_trace(go.Pie(labels=grouped[grouped['program_class'] == 'MOVIE']['hd'],
                     values=grouped[grouped['program_class'] == 'MOVIE']['No of Users who Watched'],
                     name='MOVIE', direction='clockwise', sort=True),
              row=1, col=2)

# Update layout with a title
fig.update_layout(title_text='Number of Users who Watched Vs Program Quality Flag (HD)')

# Show the plot
fig.show()

In [30]:
import plotly.express as px

# Create a new column combining 'hd' and 'program_class' for clearer labels
# Convert 'hd' to string to avoid TypeError
grouped['hd_program_class'] = grouped['program_class'] + ' (' + grouped['hd'].astype(str) + ')'

# Define a color palette for a more professional look
colors = ['#003f5c', '#58508d', '#bc5090', '#ff6361', '#ffa600']  # Professional color palette

# Create pie chart using the new combined column for names
fig_hd_program = px.pie(grouped,
                        values='No of Users who Watched',
                        names='hd_program_class',
                        hover_data=['program_class'],
                        title='HD vs SD by Program Class: Number of Users',
                        color_discrete_sequence=colors)  # Apply the professional color palette

# Update the traces to show percentage and label inside the pie chart
fig_hd_program.update_traces(textinfo='percent+label',
                             textposition='inside',
                             showlegend=True)

# Customize layout for better readability
fig_hd_program.update_layout(
    title_font_size=18,
    title_x=0.5,  # Center the title
    annotations=[dict(text='HD/SD by Program Class', x=0.5, y=-0.1, showarrow=False)],
    legend_title="Program Class (HD/SD)",
    legend=dict(
        x=1,  # Position the legend on the right side
        y=1,
        bgcolor='rgba(255, 255, 255, 0.5)',
        bordercolor='Black',
        borderwidth=1
    )
)

# Show the updated pie chart
fig_hd_program.show()


## Analysis Results
## Number of Users Who Watched by HD/SD and Program Class

| hd_program_class       | No of Users who Watched |
|------------------------|--------------------------|
| SERIES/EPISODES (0)   | 14.3%                    |
| SERIES/EPISODES (1)   | 11.5%                    |
| MOVIE (1)              | 47.6%                    |
| MOVIE (0)              | 26.6%                    |




###Conclusions from the Table:
1- **Distribution of Viewers by Quality (HD vs SD):**

- **HD Quality Content:**

- Movies: Received 47.6% of the total viewers who watched content in HD.
- Series/Episodes: Only received 11.5% of the total viewers who watched content in HD.
- **SD Quality Content**:
- Movies: Received 26.6% of the total viewers who watched content in SD.
- Series/Episodes: Received 14.3% of the total viewers who watched content in SD.

2- **Higher Preference for HD Quality:**

There is a notable preference for watching movies in HD compared to Series/Episodes. The proportion of viewers for Movies in HD (47.6%) is significantly higher than that for Series/Episodes in HD (11.5%).

3- **Contrast Between HD and SD Quality:**

Movies attract more viewers in HD compared to SD quality. Conversely, Series/Episodes are watched more in SD than HD, with Series/Episodes in SD (14.3%) being preferred over Movies in SD (26.6%).

4- **Comparison Between Quality and Program Class:**

- Movies: Dominate in HD viewership with 47.6%, indicating a strong preference for high-quality visuals when watching movies.
- Series/Episodes: Have a lower preference for HD quality, suggesting that viewers may not consider high resolution as necessary for Series/Episodes compared to Movies.
##Summary:
The data suggests that Movies are significantly preferred in HD quality compared to Series/Episodes, with a higher percentage of viewers opting for Movies in HD. On the other hand, Series/Episodes are more frequently watched in SD quality compared to Movies. This indicates a variance in viewer preferences based on content type and quality, with Movies benefiting more from high-definition quality to attract viewers.

## How does the average watch time for content in Standard Definition (SD) compare to that for High Definition (HD) resolution, and what might be the factors contributing to the observed differences in viewing duration?


In [31]:

# Group the data by resolution (HD or SD) and calculate average watch time
grouped_by_resolution = dataframe.groupby('hd').agg({
    'duration_seconds': ['mean']
}).reset_index()

# Rename columns for clarity
grouped_by_resolution.columns = ['Resolution', 'Average Watch Time (Seconds)']

# Map resolution values to more readable labels
resolution_labels = {0: 'SD', 1: 'HD'}
grouped_by_resolution['Resolution'] = grouped_by_resolution['Resolution'].map(resolution_labels)

# Plot the average watch time by resolution
fig_average_watch_time = px.bar(grouped_by_resolution, x='Resolution', y='Average Watch Time (Seconds)',
                                title='Average Watch Time by Video Resolution',
                                labels={'Average Watch Time (Seconds)': 'Average Watch Time (Seconds)'},
                                color='Resolution',
                                color_discrete_sequence=['#1f77b4', '#ff7f0e'])  # Using more serious colors

# Display the plot
fig_average_watch_time.show()

In [32]:
# Group the data by resolution (HD or SD) and calculate average watch time
grouped_by_resolution = dataframe.groupby('hd').agg({
    'duration_seconds': ['mean']
}).reset_index()

# Rename columns for clarity
grouped_by_resolution.columns = ['Resolution', 'Average Watch Time (Seconds)']

# Map resolution values to more readable labels
resolution_labels = {0: 'SD', 1: 'HD'}
grouped_by_resolution['Resolution'] = grouped_by_resolution['Resolution'].map(resolution_labels)

# Print the average watch time by resolution
print(grouped_by_resolution)

  Resolution  Average Watch Time (Seconds)
0         SD                   1501.248164
1         HD                    801.507069



### Analysis of Average Watch Time by Resolution
The analysis was performed to understand how average watch time differs between SD (Standard Definition) and HD (High Definition) resolutions. The results are summarized as follows:

**Results:**
- SD Resolution:Average Watch Time: 1501.25 seconds
- HD Resolution:Average Watch Time: 801.51 seconds

#### Interpretation:
1. Higher Average Watch Time for SD:

- The average watch time for SD resolution is significantly higher than that for HD resolution. Specifically, users spend an average of 1501.25 seconds watching content in SD compared to 801.51 seconds in HD.
- This suggests that users might prefer watching content in SD for longer periods, or there might be differences in content availability or user behavior that impact viewing time.
2. Possible Reasons for Higher SD Watch Time:

- **Content Availability**: There might be more content available in SD resolution, leading users to spend more time watching it.
- **User Preferences**: Some users may have a preference for SD resolution due to factors such as device compatibility or internet bandwidth limitations.
- **Content Type**: The type of content available in SD versus HD may differ, potentially affecting the total watch time.

3. Implications:

- **Content Strategy**: Content providers might need to consider the distribution of content between SD and HD formats based on user preferences and viewing behavior.
- **Quality vs. Quantity**: While HD provides better visual quality, SD content seems to engage users for longer durations, indicating that the quantity or type of content might be more influential than the resolution itself.

 **Conclusion**:
The analysis reveals that while HD resolution provides higher visual quality, the average watch time is greater for SD content. This highlights the importance of understanding user preferences and content availability when evaluating viewership patterns. Further investigation into content types and user behavior could provide deeper insights into these viewing trends.

